vm/devices/user_driver/src/emulated.rs

```diff
@@ -263,6 +263,25 @@ impl HostDmaAllocator for EmulatedDmaAllocator {
         memory.as_slice().atomic_fill(0);
         Ok(memory)
     }
+
+    fn attach_dma_buffer(&self, len: usize, _pfns: &[u64]) -> anyhow::Result<MemoryBlock> {
+        // For emulated allocator (unit tests) reuse the regular alloc.
+        self.allocate_dma_buffer(len)
+    }
+}
+
+#[cfg(target_os = "linux")]
+#[cfg(feature = "vfio")]
+impl crate::vfio::VfioDmaBuffer for EmulatedDmaAllocator {
+    fn create_dma_buffer(&self, len: usize) -> anyhow::Result<MemoryBlock> {
+        Ok(MemoryBlock::new(
+            self.shared_mem.alloc(len).context("out of memory")?,
+        ))
+    }
+
+    fn restore_dma_buffer(&self, len: usize, _pfns: &[u64]) -> anyhow::Result<MemoryBlock> {
+        self.create_dma_buffer(len)
```

Comment:
Same as below, you have to fail this operation.

Reply:
This didn't get fixed. You have to either do the restore faithfully or fail. If the unit tests are passing and using this function then it's because the tests are deterministic and always allocate in the same order...

Reply:
Added panic for real this time